File: hashing.py
# hashing.py

# -*- coding: utf-8 -*-

"""
Cache key hashing: SHA256 of model_id + token IDs, with word-block LCP matching.

Blocks of 100 tokens, LCP computed over full SHA256 hashes.
Key = sha256(canonical_name + '\n' + ','.join(token_ids)), i.e. model is included in the key.

Meta files contain:
- key
- model_id
- n_tokens
- wpb
- blocks
- cache_size
- recompute_penalty
"""

import hashlib
import logging
from typing import List

from config import WORDS_PER_BLOCK

META_SUFFIX = ".meta.json"

log = logging.getLogger(__name__)


def sanitize_backend_dir(backend_key: str) -> str:
    """Sanitize backend key for use as a filesystem directory name."""
    return backend_key.replace(":", "-")


def block_hashes_from_tokens(token_ids: List[int], wpb: int = WORDS_PER_BLOCK) -> List[str]:
    hashes: List[str] = []
    for i in range(0, len(token_ids), wpb):
        chunk = token_ids[i:i + wpb]
        h = hashlib.sha256(",".join(str(t) for t in chunk).encode("utf-8")).hexdigest()
        hashes.append(h)
    log.warn("Block hashes: %d blocks, %d tokens per block", len(hashes), wpb)
    return hashes


def lcp_blocks(blocks1: List[str], blocks2: List[str]) -> int:
    n = min(len(blocks1), len(blocks2))
    i = 0
    while i < n and blocks1[i] == blocks2[i]:
        i += 1
    return i


def prefix_key_sha256(text: str) -> str:
    """Basic SHA256 wrapper."""
    return hashlib.sha256(text.encode("utf-8")).hexdigest()


def meta_key(canonical_name: str, token_ids: List[int]) -> str:
    """sha256(canonical_name + '\n' + ','.join(token_ids))"""
    return prefix_key_sha256(canonical_name + "\n" + ",".join(str(t) for t in token_ids))

File: go-proxycache/hashing.go
// hashing.go — cache key hashing: SHA256 of model name + token IDs,
// with word-block LCP matching over full SHA256 block hashes.

package main

import (
	"crypto/sha256"
	"encoding/hex"
	"strconv"
	"strings"
)

const MetaSuffix = ".meta.json"

// SanitizeBackendDir converts a backend key to a safe filesystem directory name
// (colons replaced with dashes: "10.0.0.1:8000" -> "10.0.0.1-8000").
func SanitizeBackendDir(backendKey string) string {
	return strings.ReplaceAll(backendKey, ":", "-")
}

// BlockHashesFromTokens splits tokens into blocks of wpb and hashes each block
// (sha256 of comma-joined token IDs). Logs a warning per call, matching the
// Python version's behavior.
func BlockHashesFromTokens(tokenIDs []int, wpb int) []string {
	hashes := make([]string, 0, (len(tokenIDs)+wpb-1)/wpb)
	for i := 0; i < len(tokenIDs); i += wpb {
		end := i + wpb
		if end > len(tokenIDs) {
			end = len(tokenIDs)
		}
		parts := make([]string, end-i)
		for j, t := range tokenIDs[i:end] {
			parts[j] = strconv.Itoa(t)
		}
		h := sha256.Sum256([]byte(strings.Join(parts, ",")))
		hashes = append(hashes, hex.EncodeToString(h[:]))
	}
	logWarn("hashing", "Block hashes: %d blocks, %d tokens per block", len(hashes), wpb)
	return hashes
}

// BlockHashesFromTokensDefault uses the configured WORDS_PER_BLOCK.
func BlockHashesFromTokensDefault(tokenIDs []int) []string {
	return BlockHashesFromTokens(tokenIDs, WordsPerBlock)
}

// LCPBlocks returns the number of leading blocks shared by two block-hash lists.
func LCPBlocks(blocks1, blocks2 []string) int {
	n := len(blocks1)
	if len(blocks2) < n {
		n = len(blocks2)
	}
	i := 0
	for i < n && blocks1[i] == blocks2[i] {
		i++
	}
	return i
}

// PrefixKeySha256 is a basic SHA256 wrapper over UTF-8 text.
func PrefixKeySha256(text string) string {
	h := sha256.Sum256([]byte(text))
	return hex.EncodeToString(h[:])
}

// MetaKey = sha256(canonical_name + '\n' + ','.join(token_ids)).
func MetaKey(canonicalName string, tokenIDs []int) string {
	parts := make([]string, len(tokenIDs))
	for i, t := range tokenIDs {
		parts[i] = strconv.Itoa(t)
	}
	return PrefixKeySha256(canonicalName + "\n" + strings.Join(parts, ","))
}
